### ISSUE TITLE
SIGSEGV when compiling regex_automata wasm32-unknown-emscripten

### ISSUE THREAD

**Issue by lopopolo** [rust-lang/rust#69364]
### Code

```console
$ rustc --crate-name regex_automata /Users/lopopolo/.cargo/registry/src/github.com-1ecc6299db9ec823/regex-automata-0.1.8/src/lib.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts --crate-type lib --emit=dep-info,metadata,link -C debuginfo=2 -C metadata=53e0301393a8dd99 -C extra-filename=-53e0301393a8dd99 --out-dir /Users/lopopolo/dev/artichoke/playground/target/wasm32-unknown-emscripten/debug/deps --target wasm32-unknown-emscripten -L dependency=/Users/lopopolo/dev/artichoke/playground/target/wasm32-unknown-emscripten/debug/deps -L dependency=/Users/lopopolo/dev/artichoke/playground/target/debug/deps --extern byteorder=/Users/lopopolo/dev/artichoke/playground/target/wasm32-unknown-emscripten/debug/deps/libbyteorder-72d386ad6d326ca8.rmeta --cap-lints allow
{"artifact":"/Users/lopopolo/dev/artichoke/playground/target/wasm32-unknown-emscripten/debug/deps/regex_automata-53e0301393a8dd99.d","emit":"dep-info"}
{"artifact":"/Users/lopopolo/dev/artichoke/playground/target/wasm32-unknown-emscripten/debug/deps/libregex_automata-53e0301393a8dd99.rmeta","emit":"metadata"}
Called function must be a pointer!
  call addrspace(16046453) void <badref>(i8* %20), !dbg !53
in function _ZN4core6option15Option$LT$T$GT$5ok_or17h536d2927739d3b4fE
LLVM ERROR: Broken function found, compilation aborted!
```

### Meta

`rustc --version --verbose`:
```
rustc 1.41.0 (5e1a79984 2020-01-27)
binary: rustc
commit-hash: 5e1a799842ba6ed4a57e91f7ab9435947482f7d8
commit-date: 2020-01-27
host: x86_64-apple-darwin
release: 1.41.0
LLVM version: 9.0
```

`emcc --version`:
```
emcc --version
emcc (Emscripten gcc/clang-like replacement) 1.39.8 (commit 1458145cf4f3db0fb548343e6acab267eef8e4ef)
Copyright (C) 2014 the Emscripten authors (see AUTHORS.txt)
This is free and open source software under the MIT license.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Release builds complete successfully.

### Error output

```
Called function must be a pointer!
  call addrspace(16046453) void <badref>(i8* %20), !dbg !53
in function _ZN4core6option15Option$LT$T$GT$5ok_or17h536d2927739d3b4fE
LLVM ERROR: Broken function found, compilation aborted!
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
{"artifact":"/Users/lopopolo/dev/artichoke/playground/target/wasm32-unknown-emscripten/debug/deps/regex_automata-53e0301393a8dd99.d","emit":"dep-info"}
{"artifact":"/Users/lopopolo/dev/artichoke/playground/target/wasm32-unknown-emscripten/debug/deps/libregex_automata-53e0301393a8dd99.rmeta","emit":"metadata"}
Called function must be a pointer!
  call addrspace(6475917) void Segmentation fault: 11
```

</p>
</details>

**Comment by lopopolo** [comment]
I created this as an internal compiler error. I'm not sure if that's right since this is a segfault and not a panic.

**Comment by pnkfelix** [comment]
triage: P-high to determine underlying issue and actual severity, and the regression point. Removing nomination. Assigning to self.

**Comment by lopopolo** [comment]
Put together a smaller reproducer:

_Cargo.toml`
```toml
[package]
name = "bisect"
version = "0.1.0"
authors = ["Ryan Lopopolo <rjl@hyperbo.la>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
ansi_term = "0.11"
```

_src/main.rs_:
```rust
fn main() {
    println!("Hello, world!");
}
```

This project segfaults `rustc` when running:

- `cargo +1.41.1 build --target wasm32-unknown-emscripten`
- `cargo +1.41.0 build --target wasm32-unknown-emscripten`
- `cargo +1.40.0 build --target wasm32-unknown-emscripten`

With this `emcc` (1.39.8-upstream):

```console
$ emcc --version
emcc (Emscripten gcc/clang-like replacement) 1.39.8 (commit 1458145cf4f3db0fb548343e6acab267eef8e4ef)
Copyright (C) 2014 the Emscripten authors (see AUTHORS.txt)
This is free and open source software under the MIT license.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

Rust 1.39 and older use the fastcomp backend which is incompatible with the default "upstream" (LLVM) backend. Downgrading emcc to `1.38.48-fastcomp` enables builds to succeed on 1.37.0, 1.38.0, and 1.39.0.

The segfault does not appear on this nightly.

```
rustc +nightly --version --verbose
rustc 1.43.0-nightly (6d69caba1 2020-02-27)
binary: rustc
commit-hash: 6d69caba110c0c2fb90180df1cbc8be5033b91d4
commit-date: 2020-02-27
host: x86_64-apple-darwin
release: 1.43.0-nightly
LLVM version: 9.0
```

**Comment by pnkfelix** [comment]
Determined via bisection over nightlies that the problem was "fixed" between  +nightly-2020-01-25 and  +nightly-2020-01-26

**Comment by pnkfelix** [comment]
Looks like this was deliberately fixed via an LLVM upgrade in PR #68525.

**Comment by pnkfelix** [comment]
And it looks like the above fix is part of the beta channel, which is imminently scheduled for release.

Closing as duplicate of  #66308 (which is fixed).
